### PR TITLE
feat: trace context traceparent header support

### LINKED ISSUE #68: Support Trace Context HTTP headers
Ref.  
- https://github.com/w3c/trace-context#trace-context-specification
- https://w3c.github.io/trace-context/ 
  - https://w3c.github.io/trace-context/#processing-model
- https://github.com/ipfs/kubo/issues/9168 + https://github.com/ipfs/kubo/pull/9180

Quick thoughts:
- Whatever we establish here, others in IPFS Ecosystem will see, and assume this is "standard best practice" – we should plan for long term
- This should be turn-key solution for everyone running boxo/gateway, not just bifrost-gateway  
- bifrost-gateway should ensure tracing info is always present
  -  If it is missing in request, bifrost-gateway would create one. ([spec](https://w3c.github.io/trace-context/#mutating-the-traceparent-field))
  - If it is present in request, we update ids in the chain.
     - For Rhea, we will be creating trace parent on LBs ([slack thread](https://filecoinproject.slack.com/archives/C04JQ9X7ENB/p1679942350966959)), but if there is no load balancer in front of bifrost-gateway, then it should initialize trace chain.

```[tasklist]
### Tasks
- [x] https://github.com/ipfs/bifrost-gateway/pull/74
- [ ]  https://github.com/filecoin-saturn/caboose/pull/87
- [ ] https://github.com/filecoin-saturn/L1-node/issues/337
```

### PR DESCRIPTION
**Pull request by hacdias**
Closes #68.

- [x] Merge https://github.com/ipfs/boxo/pull/256
- [x] Update Boxo here

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -94,6 +97,13 @@ See documentation at: https://github.com/ipfs/bifrost-gateway/#readme`,
 			log.Fatalf("Unable to start. bifrost-gateway requires either PROXY_GATEWAY_URL or STRN_ORCHESTRATOR_URL to be set.\n\nRead docs at https://github.com/ipfs/bifrost-gateway/blob/main/docs/environment-variables.md\n\n")
 		}
 
+		// TODO: create trace provider with correct exporters and all information
+		// we may need.
+		tp := trace.NewTracerProvider()
```
Comment (hacdias) [inline main.go:102]: Kubo has a very intricate trace provider creation. Not really sure what we want to do here. This is just the default, I think.

Comment (lidel) [reply main.go:102]: @hacdias I believe we want the same thing – enabling additional trace reporters as opt-in via env vars.
We want Jaeger so we can test with Jaeger UI, but users will benefit from having other options available.  

Perhaps copy tracing.go from Kubo. All the tracing options from https://github.com/ipfs/kubo/blob/master/docs/environment-variables.md#tracing should be added to https://github.com/ipfs/bifrost-gateway/blob/main/docs/environment-variables.md.

**Review by lidel — COMMENTED** [review]
Thank you @hacdias, left a comment above.

I think the next steps here is to make this universal feature:
- [x] support what Kubo supports ([docs](https://github.com/ipfs/kubo/blob/master/docs/environment-variables.md#tracing))
- [ ] confirm that `traceparent` is applied to all requests sent by bifrost-gateway to other systems
  - [x] `KUBO_RPC_URL`  → routing.go
  - [x] when `PROXY_GATEWAY_URL`  is set  → blockstore_proxy.go
  - [ ] when `STRN_ORCHESTRATOR_URL` is set → open PR in [caboose](https://github.com/filecoin-saturn/caboose) that ensures header is forwarded to Saturn L1 with every request?
 		- This should work as per https://github.com/ipfs/bifrost-gateway/pull/74#issuecomment-1499138847
- [ ] add "How to use tracing?" to FAQ in README.md  
  - [x] example how to pass `traceparent`
  - [x] example how to trace with Jaeger UI (link to `/docs/environment-variables.md#tracing` based on Kubo?)

**Comment by hacdias** [comment]
Connected traces of Bifrost Gateway proxying requests to Kubo: 

image

**Pull request merged**